### DISCUSSION TITLE
Read access token from file

### DISCUSSION THREAD

**Discussion by enricojonas** [renovatebot/renovate#30542] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Github

### Please tell us more about your question or problem

Hi,

Currently we use RenovateBot running on GKE against Google Artifact Registry with the following config, example for Docker:

```
    "hostRules": [
        {
            "hostType": "docker",
            "matchHost": "europe-west4-docker.pkg.dev",
            "username": "_json_key_base64",
            "password": process.env.ARTIFACT_REGISTRY_SA_KEY
        }
    ]
```

Basically an exported service account key read from environment variable. As we have the directive of getting rid of all service account keys, we introduced a max lifetime of 3 months. Now we could of course implement some more or less automated key rotation solution but ideally we want to get rid of the SA keys.

As far as I know, RenovateBot does not support using workload identity on GCP. However I found out that you can also use an access token for access GAR, something like this

```
    "hostRules": [
        {
            "hostType": "docker",
            "matchHost": "europe-west4-docker.pkg.dev",
            "token": "access-token"
        }
    ]
```

As this token is only valid for 60 minutes, using environment variable isn't really an option. Plus we have the token already mounted on the filesystem by GKE (and rotated automatically).

Is it possible to read anything from file instead of environment variable?

```
    "hostRules": [
        {
            "hostType": "docker",
            "matchHost": "europe-west4-docker.pkg.dev",
            "token": "file(/path/to/file)"
        }
    ]
```

An alternative solution would be to modify the config file by a cronjob and inject the token based on expiration time but it would be nice to read directly from file.

Another question would be if config files are reloaded periodically or when changed or only on startup?

Thank you

**Comment by viceice** [comment]
use `fs.readfilesync` to read the file from `config.js`

**Comment by enricojonas** [comment]
@viceice tried it and it does not work...

```
        {
            "hostType": "docker",
            "matchHost": "europe-west4-docker.pkg.dev",
            "token": fs.readfilesync("/path/to/file")
        }
```

```
"Could not parse config file 
 TypeError: fs.readfilesync is not a function
    at Object.<anonymous> (/usr/src/app/config.js:64:25)
    at Module._compile (node:internal/modules/cjs/loader:1256:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1310:10)
    at Module.load (node:internal/modules/cjs/loader:1119:32)
    at Function.Module._load (node:internal/modules/cjs/loader:960:12)
    at Module.require (node:internal/modules/cjs/loader:1143:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at /usr/src/app/node_modules/renovate/lib/workers/global/config/parse/file.ts:25:42
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at getParsedContent (/usr/src/app/node_modules/renovate/lib/workers/global/config/parse/file.ts:25:25)
    at Object.getConfig (/usr/src/app/node_modules/renovate/lib/workers/global/config/parse/file.ts:57:14)
    at parseConfigs (/usr/src/app/node_modules/renovate/lib/workers/global/config/parse/index.ts:24:22)
    at /usr/src/app/node_modules/renovate/lib/workers/global/index.ts:125:16
    at start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:123:5)
    at /usr/src/app/node_modules/renovate/lib/renovate.ts:18:22"
```

**Reply by rarkins** [reply]
fs.readFileSync()

**Reply by viceice** [reply]
https://nodejs.org/api/fs.html#fsreadfilesyncpath-options

**Reply by enricojonas** [reply]
@rarkins obviously, I just copy/pasted before. It's working now - though I get an error as the token isn't the token that is required but it's only of the gke metadata server and can be used to obtain the access token. I'll need to write a small script that can be run as a sidecar to obtain the actual access token.

Last question, if this token is read from a file, will it be read once on renovate startup or it will be read everytime it's needed so we can update it on the fly?

**Reply by rarkins** [reply]
Once on startup

**Reply by enricojonas** [reply]
So then, this wouldn't work either. Would injecting the token into the config work? Would renovate re-read the config when it is modified while renovate is running?

**Reply by viceice** [reply]
no

**Reply by enricojonas** [reply]
So then I guess even if we would have this token in environment variable it will not pick up the new value if modified... Would it make sense to have a functionality in RenovateBot to monitor the config file and re-read it upon changes? It would be useful for such use cases, e.g. short lived tokens.

Or maybe an endpoint to trigger this reload would also be an option.
